### DISCUSSION TITLE
close window and continue with python code

### DISCUSSION THREAD

**Discussion by giannoko** [pyvista/pyvista#3489] [Q&A] [answered]
Hi,

I want to add an "add_checkbox_button_widget" to close the window and continue with the python code

The code is this 

```
def check(flag):
	print("Close the window")
	p.close()

p.add_checkbox_button_widget(check, value=True)

p.show()

print("ok")

```

when I press the check box the window is closed but the code does not continue from p.show() onwards ( "ok" is not printed)

Any help would be appriciated.

Thank you

**Comment by banesullivan** [comment]
Oof, this isn't working well. When I run this code, it segfaults/kills my kernel on the `p.close()` call within the callback. I suspect it's not safe to perform this call in the callback thread triggered by the VTK event.

We will have to more seriously investigate this.

**Comment by banesullivan** [comment, marked as answer]
There may be a workaround though by tricking the render window interaction into thinking the `q` key was pressed.

**Reply by banesullivan** [reply]
I was hoping something like this might work:

```py
def check(flag):
    p.iren.interactor.SetKeySym('q')
    p.iren.interactor.KeyReleaseEvent()
``` 

But it's not working for me... maybe this sends you in the right direction... hope this helps!

**Comment by giannoko** [comment]
Hi @banesullivan and thank you for replying. Yes it works for me! Thank you so much

**Reply by banesullivan** [reply]
Oh huh, didn't think that was working for me but glad it works for you!

**Reply by giannoko** [reply]
@banesullivan Thanks a lot!

**Reply by jesse-gohl** [reply]
This is very similar to something I was trying to do too.  I want to catch the user's "close" request then perform some additional checking before the window is actually closed and warn the user to continue until some condition is met.  The interesting thing is that my attached code works as expected with pyvista 0.36.0 but not 0.36.1 nor 0.37.0.  I'm attaching my code in case it is useful for investigating (calling the selectFacets.py function).
[Python.zip](https://github.com/pyvista/pyvista/files/10377455/Python.zip)
This is the expected output:
    state = True
    Pyvista Version = 0.36.0
    Scale = 1.7272069317100354
    Number of points = 2
In the newer versions I only get the first line, which occurs before the call to close the plotter.  Nothing after that.

**Reply by banesullivan** [reply]
@jesse-gohl, I'm pretty sure you want `auto_close=False` in your `show()` call. (Just from a quick glance... did not actually run it)

**Reply by jesse-gohl** [reply]
@banesullivan I was fairly certain I had already tried that so I went back to confirm.  Here are the results from my test:

C:\working>python -m venv myenv
C:\working>myenv\Scripts\activate.bat
(myenv) C:\working>pip install pyvista
(myenv) C:\working>pip show vtk pyvista
Name: vtk
Version: 9.2.5
Summary: VTK is an open-source toolkit for 3D computer graphics, image processing, and visualization
Home-page: https://vtk.org
Author: VTK developers
Author-email: None
License: BSD
Location: c:\working\myenv\lib\site-packages
Requires: matplotlib
Required-by: pyvista

Name: pyvista
Version: 0.37.0
Summary: Easier Pythonic interface to VTK
Home-page: https://github.com/pyvista/pyvista
Author: PyVista Developers
Author-email: info@pyvista.org
License: MIT
Location: c:\working\myenv\lib\site-packages
Requires: pillow, numpy, typing-extensions, imageio, vtk, pooch, scooby
Required-by:

(myenv) C:\working>python selectFacets.py
vtk Version = 9.2.5
pyvista Version = 0.37.0
Processing mesh...
state = True

(myenv) C:\working>

[Python.zip](https://github.com/pyvista/pyvista/files/10380679/Python.zip)

**Reply by jesse-gohl** [reply]
So, it seems I was mistaken, the issue is with the VTK version, not pyvista.  Here is the result using pyvista 0.37.0 but vtk 9.0.1:
C:\working>python -m venv myenv
C:\working>myenv\Scripts\activate.bat
(myenv) C:\working>pip install pyvista==0.36.0
(myenv) C:\working>pip show vtk pyvista
Name: vtk
Version: 9.1.0
Summary: VTK is an open-source toolkit for 3D computer graphics, image processing, and visualization
Home-page: https://vtk.org
Author: VTK developers
Author-email: None
License: BSD
Location: c:\working\myenv\lib\site-packages
Requires: wslink, matplotlib
Required-by: pyvista

Name: pyvista
Version: 0.36.0
Summary: Easier Pythonic interface to VTK
Home-page: https://github.com/pyvista/pyvista
Author: PyVista Developers
Author-email: info@pyvista.org
License: MIT
Location: c:\working\myenv\lib\site-packages
Requires: vtk, scooby, appdirs, pillow, numpy, imageio
Required-by:

(myenv) C:\working>pip install --upgrade pyvista
(myenv) C:\working>pip show vtk pyvista
Name: vtk
Version: 9.1.0
Summary: VTK is an open-source toolkit for 3D computer graphics, image processing, and visualization
Home-page: https://vtk.org
Author: VTK developers
Author-email: None
License: BSD
Location: c:\working\myenv\lib\site-packages
Requires: matplotlib, wslink
Required-by: pyvista

Name: pyvista
Version: 0.37.0
Summary: Easier Pythonic interface to VTK
Home-page: https://github.com/pyvista/pyvista
Author: PyVista Developers
Author-email: info@pyvista.org
License: MIT
Location: c:\working\myenv\lib\site-packages
Requires: pillow, pooch, vtk, scooby, numpy, imageio, typing-extensions
Required-by:

(myenv) C:\working>python selectFacets.py
vtk Version = 9.1.0
pyvista Version = 0.37.0
Processing mesh...
state = True
Finished processing mesh.
Confirm pyvista Version = 0.37.0
Scale = 1.7272069317100354
Number of points = 3

(myenv) C:\working>
